pkg/kubectl/cmd/run.go

```diff
@@ -372,81 +375,75 @@ func contains(resourcesList map[string]*unversioned.APIResourceList, resource un
 
 // waitForPod watches the given pod until the exitCondition is true. Each two seconds
 // the tick function is called e.g. for progress output.
-func waitForPod(podClient coreclient.PodsGetter, ns, name string, exitCondition func(*api.Pod) bool, tick func(*api.Pod)) (*api.Pod, error) {
-	pod, err := podClient.Pods(ns).Get(name)
+func waitForPod(podClient coreclient.PodsGetter, ns, name string, exitCondition watch.ConditionFunc, tick func(*api.Pod)) (*api.Pod, error) {
+	w, err := podClient.Pods(ns).Watch(api.SingleObject(api.ObjectMeta{Name: name}))
 	if err != nil {
 		return nil, err
 	}
-	if exitCondition(pod) {
-		return pod, nil
-	}
 
-	tick(pod)
+	pods := make(chan *api.Pod) // observed pods passed to the exitCondition
+	defer close(pods)
 
-	w, err := podClient.Pods(ns).Watch(api.SingleObject(api.ObjectMeta{Name: pod.Name, ResourceVersion: pod.ResourceVersion}))
-	if err != nil {
-		return nil, err
-	}
-
-	t := time.NewTicker(2 * time.Second)
-	defer t.Stop()
+	// wait for the first event, then start the 2 sec ticker and loop
 	go func() {
-		for range t.C {
-			tick(pod)
+		pod := <-pods
+		if pod == nil {
+			return
 		}
-	}()
+		tick(pod)
+
+		t := time.NewTicker(2 * time.Second)
+		defer t.Stop()
 
-	err = nil
-	result := pod
-	kubectl.WatchLoop(w, func(ev watch.Event) error {
-		switch ev.Type {
-		case watch.Added, watch.Modified:
-			pod = ev.Object.(*api.Pod)
-			if exitCondition(pod) {
-				result = pod
-				w.Stop()
+		for {
+			select {
+			case pod = <-pods:
+				if pod == nil {
+					return
+				}
+			case _, ok := <-t.C:
+				if !ok {
+					return
+				}
+				tick(pod)
 			}
-		case watch.Deleted:
-			w.Stop()
-		case watch.Error:
-			result = nil
-			err = fmt.Errorf("failed to watch pod %s/%s", ns, name)
-			w.Stop()
 		}
-		return nil
-	})
+	}()
 
+	intr := interrupt.New(nil, w.Stop)
+	defer intr.Close()
+	var result *api.Pod
+	intr.Run(func() error {
+		ev, err := watch.Until(0, w, func(ev watch.Event) (bool, error) {
+			c, err := exitCondition(ev)
+			if c == false && err == nil {
+				pods <- ev.Object.(*api.Pod) // send to ticker
+			}
+			return c, err
+		})
+		result = ev.Object.(*api.Pod)
+		return err
+	})
 	return result, err
 }
 
 func waitForPodRunning(podClient coreclient.PodsGetter, ns, name string, out io.Writer, quiet bool) (*api.Pod, error) {
-	exitCondition := func(pod *api.Pod) bool {
-		switch pod.Status.Phase {
-		case api.PodRunning:
-			for _, status := range pod.Status.ContainerStatuses {
-				if !status.Ready {
-					return false
-				}
-			}
-			return true
-		case api.PodSucceeded, api.PodFailed:
-			return true
-		default:
-			return false
-		}
-	}
-	return waitForPod(podClient, ns, name, exitCondition, func(pod *api.Pod) {
+	pod, err := waitForPod(podClient, ns, name, conditions.PodRunningAndReady, func(pod *api.Pod) {
```

Comment:
Looking at this, if it gets a "deleted" watch event, prior to your changes, `waitForPod` would  return the pod from the most recent "add" or "update" watch event, and no error. With your changes, it looks like it will return a "not found" error. What you're doing is probably better, but I want to make sure we think through these changes...